Newsletters: add 103 (2020-06-24) (#425)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2020-06-24-newsletter.md

---
title: 'Bitcoin Optech Newsletter #103'
permalink: /en/newsletters/2020/06/24/
name: 2020-06-24-newsletter
slug: 2020-06-24-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter summarizes a newly published fee ransom attack against
LN users, links to continued discussion about the attack against LN
atomicity, and shares a reminder about collision attacks on
RIPEMD160-based addresses in multiparty protocols.  Also included are
our regular sections with popular questions and answers from the Bitcoin
StackExchange, a list of releases and release candidates published this
week, and notable changes to popular Bitcoin infrastructure projects.

## Action items

*None this week.*

## News

- **LN fee ransom attack:** René Pickhardt [publicly
  disclosed][pickhardt post] a vulnerability to the Lightning-Dev
  mailing list that he had previously privately disclosed to LN
  implementation maintainers almost a year ago.  In the current LN
  protocol, the party who initiates opening a channel must commit to

>   paying any onchain transaction fees that will be required to later
>   close that channel.  The party paying the fees also gets to choose the

Reviewer: nit - split infinitive and extra words. Maybe simplify with something like:
```suggestion
  paying any onchain transaction fees required to close that channel.
  The party paying the fees also gets to choose the
```

  feerate used---but the security of both parties depends on choosing an

>   appropriate feerate, so the counterparty may close the channel if they

Reviewer 2: I think this should be 'refuse to open the channel' rather than 'close the channel'. BOLT 2 says "The receiving node MUST fail the channel if [...] it considers feerate_per_kw too small for timely processing or unreasonably large."

Author: That's a good point and I'll try to rephrase to be more clear, but I think the issue is more with variable feerates.  If Alice opens a channel to Bob when 10 nBTC/vbyte is sufficient to get a commitment transaction confirmed quickly and Alice agrees to pay 50 nBTC/vbyte (the 5x example), the channel will be opened and Alice can start making payments.  But if feerates then rise to 30 nBTC/vbyte being necessary, Bob's node has to consider publishing the commitement before the the safety margin of (50 - 30 = 20 nBTC/vbyte) disappears.  See [this section of BOLT2](https://github.com/lightningnetwork/lightning-rfc/blob/master/02-peer-protocol.md#updating-fees-update_fee) for details (which is the section where the 5x example comes from).

    If that feerate is five times higher than strictly necessary, this
    can easily result in paying more than a hundred dollars in
    transaction fees.  Additionally, if the commitment transaction is
    confirmed, the HTLCs then need to be settled (again using
    time-sensitive transactions that may need to pay a high feerate).
    If the victim was the party routing those payments outbound, they'll
    need to pay an additional transaction fee to recover each payment,
    which could make the attack two or three times more costly for them.

    Fees are paid to miners, so there's no direct motivation to perform
    this attack, but if the attacker has the means to quickly contact
    the victim, it's possible they can offer to settle the HTLCs
    offchain in return for a ransom, allowing the attacker to profit.

    Pickhardt concludes his email with several ideas for addressing the
    problem, none of which he finds completely satisfactory.  The
    mitigation originally implemented in Eclair and later implemented in
    C-Lightning (see [Newsletter #59][news59 cl30]) is for LN nodes to limit the
    number of pending payments, keeping transactions small and total

>     fees low.  Another mitigation in development are [anchor
>     outputs][topic anchor outputs] which allow feerates to be selected

Reviewer: ```suggestion
    fees low.  Another mitigation in development is [anchor
    outputs][topic anchor outputs], which allow feerates to be selected
```

    when the channel is closed---eliminating the need to overestimate
    fees in order to prevent premature channel closures.  Several other
    ideas are mentioned, but Pickhardt asks readers to contemplate the
    problem and suggest any other possible solutions.

- **Continued discussion about LN atomicity attack:** Bastien Teinturier
  [posted][teinturier post] to the Lightning-Dev mailing list with a
  link to a [detailed description][teinturier gist] of the LN commitment

>   protocol, it's weaknesses, and proposals to address those weaknesses.

Reviewer 2: s/it's/its/

  The document teaches readers everything they should need to know to
  understand the attack against LN atomicity described in [Newsletter
  #95][news95 ln attack] as well as several proposed mitigations.  The
  document's clear writing tied together several fragmented threads from

>   previous discussions.  This lead to renewed evaluation of several

Reviewer: ```suggestion
  previous discussions.  This led to a renewed evaluation of several
```

  LND maintenance release is now available for testing.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

- [Bitcoin Core #19260][] p2p: disconnect peers that send filterclear + update existing filter msg disconnect logic FIXME:dongcarl

- [Bitcoin Core #19133][] rpc, cli, test: add bitcoin-cli -generate command FIXME:moneyball

- [Bitcoin Core #18027][] adds two options to the GUI's *File* menu for
  working with Partially-Signed Bitcoin Transactions ([PSBTs][topic
  psbt]): *Load PSBT from file* and *Load PSBT from clipboard*.  When
  one of those options is clicked and a PSBT is loaded, a dialog is
  provided that allows the user to sign an incomplete PSBT if this
  wallet has its key, broadcast a complete PSBT, or copy or save the
  PSBT for processing with another tool such as [HWI][].  Combined with
  other recent PSBT-related GUI changes (see Newsletters [#74][news74

>   psbtgui] and [#82][news82 psbtgui]) and [HWI's own GUI][hwi gui] this

Reviewer: ```suggestion
  psbtgui] and [#82][news82 psbtgui]) and [HWI's own GUI][hwi gui], this
```

  makes it possible to use a PSBT-based process with Bitcoin Core for
  the first time without using any RPCs.

    ![Screenshot of PSBT dialog in Bitcoin Core GUI](/img/posts/2020-06-psbt-dialog.png)

- [Bitcoin Core #16377][] updates the `walletcreatefundedpsbt` and
  `fundrawtransaction` RPCs.  These RPCs normally use the wallet to

>   automatically choose which UTXOs to spend in a unsigned transaction,

Reviewer: ```suggestion
 choose which UTXOs to spend in an unsigned transaction automatically,
```

Author: What's the motivation for moving "automatically"?  (I'm not opposed, just curious.)

Reviewer: It's a split infinitive.

  but they also allow the user to specify one or more UTXOs they want to
  spend in that transaction.  Previously, if the UTXOs selected by the
  user weren't enough to pay for all of the transaction's outputs, the
  wallet would automatically select more UTXOs to spend.  But, if a
  user is manually selecting UTXOs, they may have some reason they don't

>   want to spend additional UTXOs, the the RPCs will now fail by default

Reviewer: ```suggestion
  want to spend additional UTXOs, the RPCs will now fail by default
```

Reviewer 3: s/the the/and the/ (or another way to link the phrases)

  this new repository, which will be bidirectionally synced with the
  [main repository's master development branch][bitcoin/bitcoin] using
  the *monotree* development model used by the Linux Kernel Project.
  There are no direct user-visible changes from this split---users will
  still receive the GUI in official packaged versions of Bitcoin Core or
  when building using `--with-gui` from the source code in the main
  repository.

    FIXME:harding illustration of monotree vs monorepo

    This split is an experiment designed to determine whether using
    different repositories for different subsystems will help people
    interested in a particular subsystem focus on that part of the
    project.  For example, someone using the GitHub *Watch Repository*
    feature will receive fewer issue and PR status updates each day when
    watching the Bitcoin Core GUI repository rather than the main
    repository, making it easier for them to monitor the project.
    Conversely, developers watching the main project, who may not all
    be interested in the GUI, will no longer need to receive
    notifications about it.  In the best case, it's hoped that this

>     improved focus can speed up development---allowing the the project to

Reviewer: ```suggestion
    improved focus can speed up development---allowing the project to
```

>     create monotree repositories for other parts of the project.

Reviewer 2: The dash makes it seem like "allowing the project to create monotree repositories for other parts of the project." is the goal. I think it should be a separate sentence to clarify that separating the project further is possible but isn't proposed right now.

Also, s/the the/the/

RIPEMD160-based addresses in multiparty protocols.  Also included are
our regular sections with popular questions and answers from the Bitcoin
StackExchange, a list of releases and release candidates published this
week, and notable changes to popular Bitcoin infrastructure projects.

## Action items

*None this week.*

## News

- **LN fee ransom attack:** René Pickhardt [publicly
  disclosed][pickhardt post] a vulnerability to the Lightning-Dev
  mailing list that he had previously privately disclosed to LN
  implementation maintainers almost a year ago.  In the current LN
  protocol, the party who initiates opening a channel must commit to
  paying any onchain transaction fees that will be required to later
  close that channel.  The party paying the fees also gets to choose the
  feerate used---but the security of both parties depends on choosing an
  appropriate feerate, so the counterparty may close the channel if they
  think the chosen feerate is too low.  To avoid such unnecessary
  closes, [BOLT2][] recommends the fee-paying party commit to fees five
  times higher than the minimum reasonable estimate---high enough to

>   ensure the counterparty will be satisfied even if they're using a

Reviewer 3: From BOLT2: *Given the variance in fees, and the fact that the transaction may be spent in the future, it's a good idea for the fee payer to keep a good margin (say 5x the expected fee requirement); but, due to differing methods of fee estimation, an exact value is not specified.*

- "ensure" and "high enough": I wonder if these words are too reassuring for describing a hopeful, non-guaranteed margin of safety (IIUC)

- "recommends": the 5x looks like more of an off-the-cuff, to-give-an-idea-of-order-of-magnitude example than a recommendation

  different fee estimation algorithm.

    BOLT2 also allows channels to route up to 483 payments
    simultaneously, each of which requires a 43 vbyte P2WSH output, for
    a total of about 20,000 vbytes of data that needs to be added to the
    chain relatively quickly---meaning it may need to pay a high feerate.
    If that feerate is five times higher than strictly necessary, this

>     can easily result in paying more than a hundred dollars in

Reviewer 3: a sum described in dollars (USD? there are many different currencies called dollars) may not age well

    outputs][topic anchor outputs] which allow feerates to be selected
    when the channel is closed---eliminating the need to overestimate
    fees in order to prevent premature channel closures.  Several other
    ideas are mentioned, but Pickhardt asks readers to contemplate the
    problem and suggest any other possible solutions.

- **Continued discussion about LN atomicity attack:** Bastien Teinturier
  [posted][teinturier post] to the Lightning-Dev mailing list with a
  link to a [detailed description][teinturier gist] of the LN commitment
  protocol, it's weaknesses, and proposals to address those weaknesses.
  The document teaches readers everything they should need to know to
  understand the attack against LN atomicity described in [Newsletter
  #95][news95 ln attack] as well as several proposed mitigations.  The
  document's clear writing tied together several fragmented threads from
  previous discussions.  This lead to renewed evaluation of several
  previously-proposed solutions, including concerns about the
  effectiveness of the "alternative anchor proposal" and a [suggested
  procedure][zmn procedure] to use the [pay-for-signature][] scriptless

>   script to trustlessly pay a third-party for the final signature needed

Reviewer 3: remove hyphens? "previously proposed" and "third party" (since the latter isn't being used as a combined adjective)

  to complete an adaptor signature.

- **Reminder about collision attack risks on two-party ECDSA:**
  cryptographer Jonas Nick [replied][nick collision] to the Bitcoin-Dev
  mailing list thread about a proposed CoinSwap implementation (see
  [Newsletter #100][news100 coinswap]) reminding developers that P2PKH,
  P2WPKH, and P2SH addresses which use the 160-bit RIPEMD160 hash are
  vulnerable to collision attacks that reduce its security to 80 bits
  when multiple parties collaborate to create an address using a naive
  protocol (see our [description of this weakness][address security]
  in legacy P2SH addresses).  Although this was previously only a
  concern to users of P2SH multisig, it applies in this context where

>   it's proposed that two users could share a P2PKH or P2WPKH address.

Reviewer 3: maybe s/it's proposed/CoinSwap proposes/

  previous transaction hash, such an attack is not feasible.

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [LND 0.10.2-beta.rc2][lnd 0.10.2-beta] this release candidate for an
  LND maintenance release is now available for testing.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

- [Bitcoin Core #19260][] p2p: disconnect peers that send filterclear + update existing filter msg disconnect logic FIXME:dongcarl

- [Bitcoin Core #19133][] restores the bitcoin-cli `generate` command
  which had been removed from Bitcoin Core. It is a shortcut for
  `generatetoaddress(getnewaddress())` and is useful for testing with
  regtest.
  
- [Bitcoin Core #18027][] adds two options to the GUI's *File* menu for

>   working with Partially-Signed Bitcoin Transactions ([PSBTs][topic

Reviewer 3: no hyphen in PSBT

  psbt]): *Load PSBT from file* and *Load PSBT from clipboard*.  When
  one of those options is clicked and a PSBT is loaded, a dialog is
  provided that allows the user to sign an incomplete PSBT if this
  wallet has its key, broadcast a complete PSBT, or copy or save the
  PSBT for processing with another tool such as [HWI][].  Combined with
  other recent PSBT-related GUI changes (see Newsletters [#74][news74
  psbtgui] and [#82][news82 psbtgui]) and [HWI's own GUI][hwi gui] this
  makes it possible to use a PSBT-based process with Bitcoin Core for
  the first time without using any RPCs.

    ![Screenshot of PSBT dialog in Bitcoin Core GUI](/img/posts/2020-06-psbt-dialog.png)

- [Bitcoin Core #16377][] updates the `walletcreatefundedpsbt` and
  `fundrawtransaction` RPCs.  These RPCs normally use the wallet to
  automatically choose which UTXOs to spend in a unsigned transaction,
  but they also allow the user to specify one or more UTXOs they want to
  spend in that transaction.  Previously, if the UTXOs selected by the
  user weren't enough to pay for all of the transaction's outputs, the

>   wallet would automatically select more UTXOs to spend.  But, if a

Reviewer 3: maybe s/But,/However,/

  user is manually selecting UTXOs, they may have some reason they don't
  want to spend additional UTXOs, the the RPCs will now fail by default
  if the user manually selects any UTXOs.  This default may be overridden
  using the new `add_inputs` parameter to both RPCs.

- [Eclair #1461][] adds several API endpoints that forward to Bitcoin

>   Core RPCs for getting that program's wallet balances and other

Reviewer 3: maybe s/getting/fetching|retrieving/ and omit "that program's"

Author: Will do s/getting/relaying/, but I'd like to keep "that program's" in order to differentiate from Eclair providing its own wallet's balance information.

  information.  The goal is to make it easier to integrate Eclair with
  the [Ride The Lightning][] node management dashboard.

- [Bitcoin Core #19071][] adds documentation describing how developers
  can contribute to the new and experimental [Bitcoin Core GUI
  repository][].  Pull requests related to the GUI should be made into
  this new repository, which will be bidirectionally synced with the
  [main repository's master development branch][bitcoin/bitcoin] using
  the *monotree* development model used by the Linux Kernel Project.
  There are no direct user-visible changes from this split---users will

>   still receive the GUI in official packaged versions of Bitcoin Core or

Reviewer 3: maybe s/in official/in the official/

  when building using `--with-gui` from the source code in the main
  repository.

    FIXME:harding illustration of monotree vs monorepo

    This split is an experiment designed to determine whether using
    different repositories for different subsystems will help people
    interested in a particular subsystem focus on that part of the
    project.  For example, someone using the GitHub *Watch Repository*
    feature will receive fewer issue and PR status updates each day when
    watching the Bitcoin Core GUI repository rather than the main
    repository, making it easier for them to monitor the project.
    Conversely, developers watching the main project, who may not all
    be interested in the GUI, will no longer need to receive
    notifications about it.  In the best case, it's hoped that this
    improved focus can speed up development---allowing the the project to
    create monotree repositories for other parts of the project.
    In the worst case, it's feared the split could slow down

>     development---but, if that happens, the experiment can be easily

Reviewer 3: maybe clarify: "slow down development by dispersing or not attracting new resources"

  Ravi Patel asks why a simpler formula for calculating the difficulty target
  from `nBits` wasn’t chosen. Andrew Chow dives into some of the details around
  the formula, its history, and even sample code from Bitcoin’s 0.1.5 release.

- [Does Bitcoin really need timestamps?]({{bse}}96185)
  Pieter Wuille explains why limiting the block rate without reference to a
  clock time outside of the blockchain could make running full nodes more
  expensive while also struggling to keep the stale block rate low and
  preventing collusion attacks.

- [In a fee overpayment attack, why can't compromised software provide fake previous transactions corresponding to fake inputs?]({{bse}}96309)
  Regarding a [fee overpayment attack][news101 fee overpayment attack] on segwit
  transactions with multiple inputs, justinmoon asks why the remedy of the
  attack, requiring copies of previous transactions for the inputs, is not
  vulnerable to malicious software providing fake previous transactions. Since
  any provided previous transaction must have a hash that matches the spending input’s
  previous transaction hash, such an attack is not feasible.

> ## Releases and release candidates

Reviewer 3: Should we mention the HWI release candidate? https://github.com/bitcoin-core/HWI/releases/tag/1.1.2-rc.2

Author: I currently only list releases and RCs from the projects listed in the notable commits section, which doesn't currently include HWI.  I had never thought about including HWI in that list until you mentioned it---I guess I never considered it when HWI was a new and experimental project, but now that it's used by several other programs, it's become a legit infrastructure project.

Let's skip an announcement for this week---it's just an RC---but does anyone have any objections to adding HWI to the list of projects we track for notable merges starting next week?

Reviewer 3: :+1: for HWI coverage.

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [LND 0.10.2-beta.rc2][lnd 0.10.2-beta] this release candidate for an
  LND maintenance release is now available for testing.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

- [Bitcoin Core #19260][] p2p: disconnect peers that send filterclear + update existing filter msg disconnect logic FIXME:dongcarl

> - [Bitcoin Core #19133][] restores the bitcoin-cli `generate` command

Reviewer 2: "restores" isn't right here, since there was never a bitcoin-cli generate command previously. More accurate would be: "Bitcoin Core 19133 adds a bitcoin-cli `generate` alias to replace the functionality of the bitcoind `generate` RPC method, which was removed from Bitcoin Core in V..."

  attack, requiring copies of previous transactions for the inputs, is not
  vulnerable to malicious software providing fake previous transactions. Since
  any provided previous transaction must have a hash that matches the spending input’s
  previous transaction hash, such an attack is not feasible.

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [LND 0.10.2-beta.rc2][lnd 0.10.2-beta] this release candidate for an
  LND maintenance release is now available for testing.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

- [Bitcoin Core #19260][] disables the overloading of `filterclear` messages for
  non-bloomfilter peers to request transaction relay commencement. Previously,

>   this behaviour was used to delay transaction relay until after IBD is done.

Reviewer 2: I suggest changing "this behaviour was used" to "this could have been used". I don't know if any clients ever actually did this (Bitcoin Core didn't).

Reviewer 4: How about: "this behaviour allowed delaying transaction relay..."?

Reviewer 2: I prefer 'could have' because I'm skeptical that any nodes actually used this behavior, and 'could have' captures that better. If there was evidence that a node was using this, then I think 'allowed delaying transaction relay' would be better.

I think both are fine though.

  user is manually selecting UTXOs, they may have some reason they don't
  want to spend additional UTXOs, so the RPCs will now fail by default
  if the user manually selects any UTXOs.  This default may be overridden
  using the new `add_inputs` parameter to both RPCs.

- [Eclair #1461][] adds several API endpoints that forward to Bitcoin
  Core RPCs for relaying that program's wallet balances and other
  information.  The goal is to make it easier to integrate Eclair with
  the [Ride The Lightning][] node management dashboard.

- [Bitcoin Core #19071][] adds documentation describing how developers
  can contribute to the new and experimental [Bitcoin Core GUI
  repository][].  Pull requests related to the GUI should be made into
  this new repository, which will be bidirectionally synced with the
  [main repository's master development branch][bitcoin/bitcoin] using
  the *monotree* development model used by the Linux Kernel Project.
  There are no direct user-visible changes from this split---users will
  still receive the GUI in the official packaged versions of Bitcoin Core or
  when building using `--with-gui` from the source code in the main
  repository.

>     ![Illustration of monorepo versus monotree](/img/posts/2020-06-monorepo-vs-monotree.png)

Reviewer 2: Using Master for the main repo is slightly confusing in this diagram, since the GUI fork will also have a master branch. Would changing Master to bitcoin/bitcoin and GUI to bitcoin-core/gui make this clearer?

Author: Good idea!  Done.

Reviewer 3: The diagram is great. It would need a colored rather than transparent background to work with dark mode.

Author: We don't natively support a dark mode, so I think it's up to whatever feature or plugin you're using to invert or otherwise mutate images to make them work with dark mode.  If there are any HTML tags or CSS changes we can add to make that possible, I'm happy to do that, but I prefer to design most of my images so that they'll look good in print---which means assuming a white background.

  #95][news95 ln attack] as well as several proposed mitigations.  The
  document's clear writing tied together several fragmented threads from
  previous discussions.  This led to a renewed evaluation of several
  previously proposed solutions, including concerns about the
  effectiveness of the "alternative anchor proposal" and a [suggested
  procedure][zmn procedure] to use the [pay-for-signature][] scriptless
  script to trustlessly pay a third party for the final signature needed
  to complete an adaptor signature.

- **Reminder about collision attack risks on two-party ECDSA:**
  cryptographer Jonas Nick [replied][nick collision] to the Bitcoin-Dev
  mailing list thread about a proposed CoinSwap implementation (see
  [Newsletter #100][news100 coinswap]) reminding developers that P2PKH,
  P2WPKH, and P2SH addresses which use the 160-bit RIPEMD160 hash are
  vulnerable to collision attacks that reduce its security to 80 bits
  when multiple parties collaborate to create an address using a naive
  protocol (see our [description of this weakness][address security]
  in legacy P2SH addresses).  Although this was previously only a

>   concern to users of P2SH multisig, it applies in contexts, such as CoinSwap, where

Reviewer 3: "a concern for" or "of concern for"

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

- [Bitcoin Core #19260][] disconnects peers that send the [BIP37][]
  `filterclear` message if the local node isn't accepting bloom filters
  (as advertised using the [BIP111][] `NODE_BLOOM` service flag).  It
  was [previously proposed][Bitcoin Core #8709] that nodes starting in
  Initial Blockchain Download (IBD) could register as non-relaying peers
  to avoid receiving recent transactions while they were still
  downloading large numbers of blocks.  When they finished syncing, they
  could then transition to receiving relayed transactions by sending a
  `filterclear` message.  However, it was more recently
  [proposed][Bitcoin Core #19204] that this could be done instead using
  the [BIP133][] `feefilter` message.  That eliminates any need for a
  non-bloom node to support the `filterclear` message and so this PR
  removes that feature.

> - [Bitcoin Core #19133][] adds a bitcoin-cli `-generate` parameter (note leading dash) to replace

Author: IMO, `bitcoin-cli` is a command and `-generate` is a parameter or argument.

> - [Bitcoin Core #19133][] adds a bitcoin-cli `-generate` parameter (note leading dash) to replace

Reviewer 3: They are commands in the help.
```
$ bitcoin-cli -help
Bitcoin Core RPC client version ...

Usage:  bitcoin-cli [options] <command> [params]  Send command to Bitcoin Core
or:     bitcoin-cli [options] -named <command> [name=value]...  Send command to Bitcoin Core (with named arguments)
or:     bitcoin-cli [options] help                List commands
or:     bitcoin-cli [options] help <command>      Get help for a command
```

In practice, of the commands and options listed afterwards in the help (in `src/bitcoin-cli.cpp`), most are options and two of them or more, `-getinfo` and `-generate` (and arguably the help/h/? help commands and `-version`), are commands.

> - [Bitcoin Core #19133][] adds a bitcoin-cli `-generate` parameter (note leading dash) to replace

Author: Nope, they're arguments in the help.  :-)  See this line specifically:

    or:     bitcoin-cli [options] -named <command> [name=value]...  Send command to Bitcoin Core (with named arguments)

What it's calling commands are RPCs.

> - [Bitcoin Core #19133][] adds a bitcoin-cli `-generate` parameter (note leading dash) to replace

Reviewer 3: I see your point, but describing -generate as a bitcoin-cli param/arg, that itself takes params/args to pass to the RPC command, seems convoluted to me. Maybe "option"? The help lists them as "Options".

Perhaps prefixing "command" with "RPC" in the help would also be clearer. :man_shrugging:

> - [Bitcoin Core #19133][] adds a bitcoin-cli `-generate` parameter (note leading dash) to replace

Reviewer 3: I've been reading that help incorrectly for years now :p

> - [Bitcoin Core #19133][] adds a bitcoin-cli `-generate` parameter (note leading dash) to replace

Author: I use "command" for the name of the program run on the *command* line.  I use "parameter" for attributes passed to a command or RPC that modify its behavior.  I try to avoid "argument" because it sometimes requires extra words to make it clear I'm not talking about an argument between people.  I try to avoid "option" unless the parameter is actually optional (which, in `bitcoin-cli`'s case, is currently all parameters).

> - [Bitcoin Core #19133][] adds a bitcoin-cli `-generate` parameter (note leading dash) to replace

Reviewer 3: Thanks @harding, as always learned much from discussing with you. (I'll probably continue to think about it as a command running inside the cli program, but I see the cli help with new eyes now :)

  the functionality of the bitcoind `generate` RPC method, which was removed
  from Bitcoin Core in [Version 0.19.0.1][core version]. The new
  implementation avoids creating unnecessary dependencies between the

>   wallet and other components. Providing an alias for the old RPC is useful for

Reviewer 3: I suggested s/alias/client-side version/ because an alias for a no-longer existing RPC might be confusing and to highlight this is not a server-side RPC that is being restored. Could be wrong, but I suspect there is some confusion regarding the distinction (and why) between the RPC API and the CLI interface.

>   wallet and other components. Providing an alias for the old RPC is useful for

Author: Updated to `client-side alias` as a compromise.  I agree that there's confusion---I think the initial description here demonstrated that!

  automatically choose which UTXOs to spend in an unsigned transaction,
  but they also allow the user to specify one or more UTXOs they want to
  spend in that transaction.  Previously, if the UTXOs selected by the
  user weren't enough to pay for all of the transaction's outputs, the
  wallet would automatically select more UTXOs to spend.  But, if a
  user is manually selecting UTXOs, they may have some reason they don't
  want to spend additional UTXOs, so the RPCs will now fail by default
  if the user manually selects any UTXOs.  This default may be overridden
  using the new `add_inputs` parameter to both RPCs.

- [Eclair #1461][] adds several API endpoints that forward to Bitcoin
  Core RPCs for relaying that program's wallet balances and other
  information.  The goal is to make it easier to integrate Eclair with
  the [Ride The Lightning][] node management dashboard.

- [Bitcoin Core #19071][] adds documentation describing how developers
  can contribute to the new and experimental [Bitcoin Core GUI
  repository][].  Pull requests related to the GUI should be made into
  this new repository, which will be bidirectionally synced with the

>   [main repository's master development branch][bitcoin/bitcoin] using

Reviewer 3: can omit "development", or as an alternative (given possible adoption): "primary repository's main branch" (as per https://www.zdnet.com/article/github-to-replace-master-with-alternative-term-to-avoid-slavery-references)

  disclosed][pickhardt post] a vulnerability to the Lightning-Dev
  mailing list that he had previously privately disclosed to LN
  implementation maintainers almost a year ago.  In the current LN
  protocol, each time the channel state is updated, the party who
  initiated opening the channel must commit to paying any onchain
  transaction fees for the unilateral close transaction.  The party
  paying the fees also gets to choose the feerate used---but both

>   party's security depends on the feerate being appropriate.  That

Reviewer 3: s/party's/parties'/ (not sure, but this looks off)

- [Bitcoin Core #19260][] p2p: disconnect peers that send filterclear + update existing filter msg disconnect logic FIXME:dongcarl

- [Bitcoin Core #19133][] restores the bitcoin-cli `generate` command
  which had been removed from Bitcoin Core. It is a shortcut for
  `generatetoaddress(getnewaddress())` and is useful for testing with

>   regtest.

Reviewer 3: `generate` used to be a server-side RPC and (according to https://github.com/bitcoin/bitcoin/issues/14299) was removed for software architectural reasons.

One reason for its return is many of the historical tutorials were broken when it was removed (per https://github.com/bitcoin/bitcoin/pull/17700#issuecomment-566159096).

It was brought back as a client-side CLI `-generate` command (denoted with a preceding dash), outside of the RPC API because that is considered the right place for CLI commands for use by humans, for example per https://github.com/bitcoin/bitcoin/pull/16439#issuecomment-516926539.

I have to write a release note for it now; will share the link here.

Reviewer 3: Release note https://github.com/bitcoin/bitcoin/pull/19357/files
